malariagen_data/anoph/dipclust.py

```diff
@@ -0,0 +1,270 @@
+        self,
+        region: base_params.regions,
+        site_mask: base_params.site_mask = DEFAULT,
+        sample_sets: Optional[base_params.sample_sets] = None,
+        sample_query: Optional[base_params.sample_query] = None,
+        cohort_size: Optional[base_params.cohort_size] = None,
+        distance_metric: dipclust_params.distance_metric = dipclust_params.distance_metric_default,
+        random_seed: base_params.random_seed = 42,
+    ) -> Tuple[np.ndarray, np.ndarray, int]:
+        # Change this name if you ever change the behaviour of this function, to
+        # invalidate any previously cached data.
+        name = "diplotype_pairwise_distances"
+
+        # Normalize params for consistent hash value.
+        sample_sets_prepped = self._prep_sample_sets_param(sample_sets=sample_sets)
+        region_prepped = self._prep_region_cache_param(region=region)
+        params = dict(
+            region=region_prepped,
+            site_mask=site_mask,
+            sample_sets=sample_sets_prepped,
+            sample_query=sample_query,
+            cohort_size=cohort_size,
+            distance_metric=distance_metric,
+            random_seed=random_seed,
+        )
+
+        # Try to retrieve results from the cache.
+        try:
+            results = self.results_cache_get(name=name, params=params)
+
+        except CacheMiss:
+            results = self._diplotype_pairwise_distances(**params)
+            self.results_cache_set(name=name, params=params, results=results)
+
+        # Unpack results")
+        dist: np.ndarray = results["dist"]
+        gt_samples: np.ndarray = results["gt_samples"]
+        n_snps: int = int(results["n_snps"][()])  # ensure scalar
+
+        return dist, gt_samples, n_snps
+
+    def _diplotype_pairwise_distances(
+        self,
+        *,
+        region,
+        site_mask,
+        sample_sets,
+        sample_query,
+        cohort_size,
+        distance_metric,
+        random_seed,
+    ):
+        import dask.array as da
+
+        if distance_metric == "cityblock":
+            metric = multiallelic_diplotype_mean_cityblock
+        elif distance_metric == "euclidean":
+            metric = multiallelic_diplotype_mean_sqeuclidean
+
+        # Load haplotypes.
+        ds_snps = self.snp_calls(
+            region=region,
+            sample_query=sample_query,
+            sample_sets=sample_sets,
+            site_mask=site_mask,
+            cohort_size=cohort_size,
+            random_seed=random_seed,
+        )
+
+        with self._dask_progress(desc="Load genotypes"):
+            gt = ds_snps["call_genotype"].data.compute()
+
+        # Compute allele count, remove non-segregating sites.
+        ac = allel.GenotypeArray(gt).count_alleles(max_allele=3)
+        gt_seg = gt.compress(ac.is_segregating(), axis=0)
+        ac_seg = allel.GenotypeArray(gt_seg).to_allele_counts(max_allele=3)
+        X = da.swapaxes(ac_seg.values, 0, 1).compute()
```

Comment:
...although if using numpy will want to make sure the result gets copied into a contiguous (C order) array, as that will go faster for pairwise distance. Memory layout makes a difference for pairwise distance because of the way the data is being accessed.

Reply:
`np.swapaxes()` did not result in a contiguous array, so just passed the result to `np.ascontiguousarray()`